### ISSUE TITLE
new lint rule - dublicate generics in `where`

### ISSUE THREAD

**Issue by ArtemIsmagilov** [rust-lang/rust-clippy#13876]
### What it does

I don't really understand which one this rule should apply to - rustfmt or clippy.

I would like duplicate generics to be highlighted with a warning or automatically formatted with deletion. It's hard for me to say what should be correct.

### Advantage

There will be no duplicate generics in `where`. 

### Drawbacks

There is not![Image](https://github.com/user-attachments/assets/41f3d66a-311e-4378-a539-301202fed112)

### Example

show error only func `new()`
```rust
fn main() {
    assert_eq!(0.0, new());
    assert_eq!(0.0, new_where());

    assert_eq!(0, new());
    assert_eq!(0, new_where());
}

fn new<T: Default, `T: Default`>() -> T {
    T::default()
}

fn new_where<T>() -> T
where
    T: Default,
    T: Default,
{
    T::default()
}
```

Could be written as:

Must be show errors in funcs `new` and `new_where`. 
```rust
fn main() {
    assert_eq!(0.0, new());
    assert_eq!(0.0, new_where());

    assert_eq!(0, new());
    assert_eq!(0, new_where());
}

fn new<T: Default, `T: Default`>() -> T {
    T::default()
}

fn new_where<T>() -> T
where
    T: Default,
    `T: Default,`
{
    T::default()
}
```

**Comment by y21** [comment]
This lint already exists: https://rust-lang.github.io/rust-clippy/master/index.html#type_repetition_in_bounds

But it's currently in the nursery due to bugs. Putting `#![warn(clippy::type_repetition_in_bounds)]` at the top I get
```
warning: this type has already been used as a bound predicate
  --> src/main.rs:12:5
   |
12 |     T: Default, // Duplicate condition, not linted
   |     ^^^^^^^^^^
   |
   = help: consider combining the bounds: `T: Default + Default`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_repetition_in_bounds

```

**Comment by samueltardieu** [comment]
> `fn new<T: Default, T: Default /*linted*/>()  // not compiled`

Here, you are declaring a function `new` with two generic arguments, `T` and `T`, with the same name (which is forbidden).

>  `fn new<T: Default + Default /*unlinted*/>() // was compiled`

Here, you are declaring a function `new` with one generic argument, `T`, that must implement `Default` as well as `Default` (redundant, but not forbidden).

**Comment by y21** [comment]
When you write `fn new<T, T>() {}`, you are introducing two *different* generic parameters with the same name `T`. That's not allowed (for the same reason why e.g. `struct S {} struct S {}` is not allowed - it's ambiguous whether the first or second type is chosen when you write `T` or `S` given that type definition order doesn't matter in the same scope). The trait bound part `: Default` is unrelated to this, the problem is that you have two generic parameters with the same name.

`fn new<T: Default + Default>` is fine because it only adds one type parameter with name `T` and adds two bounds.

For where clauses, there is no difference between `where T: Default, T: Default` and `where T: Default + Default`, but the former is preferred, so it gets linted.

**Comment by ArtemIsmagilov** [comment]
Thank you very much for opening your eyes to these subtleties. Now it’s clear what the differences are!
